### ISSUE TITLE
`Given a` machine does not run apt-get update on it

### ISSUE THREAD

**Issue by panlinux** [canonical/ubuntu-pro-client#2859]
**Description of the bug**

Tests where an extra machine is created, and where packages are installed via apt-get, can fail to install the packages if the repository information is out of date. I'm seeing this in the [ubuntu_pro.feature](https://github.com/canonical/ubuntu-pro-client/actions/runs/7023435516/job/19110296166?pr=2794) set of tests where a proxy machine is created, and the step to install squid fails on it:
```
The following NEW packages will be installed:
        libdbi-perl libecap3 squid squid-common squid-langpack ssl-cert
      0 upgraded, 6 newly installed, 0 to remove and 0 not upgraded.
      Need to get 3687 kB of archives.
      After this operation, 15.7 MB of additional disk space will be used.
      Get:1 http://us-west2.gce.archive.ubuntu.com/ubuntu focal/main amd64 libecap3 amd64 1.0.1-3.2ubuntu1 [17.4 kB]
      Get:2 http://us-west2.gce.archive.ubuntu.com/ubuntu focal/main amd64 squid-langpack all 20191103-1 [167 kB]
      Ign:3 http://us-west2.gce.archive.ubuntu.com/ubuntu focal-updates/main amd64 squid-common all 4.10-1ubuntu1.7
      Get:4 http://us-west2.gce.archive.ubuntu.com/ubuntu focal-updates/main amd64 libdbi-perl amd64 1.643-1ubuntu0.1 [730 kB]
      Get:5 http://us-west2.gce.archive.ubuntu.com/ubuntu focal/main amd64 ssl-cert all 1.0.39 [17.0 kB]
      Ign:6 http://us-west2.gce.archive.ubuntu.com/ubuntu focal-updates/main amd64 squid amd64 4.10-1ubuntu1.7
      Err:3 http://security.ubuntu.com/ubuntu focal-updates/main amd64 squid-common all 4.10-1ubuntu1.7
        404  Not Found [IP: 104.198.11.147 80]
      Err:6 http://security.ubuntu.com/ubuntu focal-updates/main amd64 squid amd64 4.10-1ubuntu1.7
        404  Not Found [IP: 104.198.11.147 80]
      Fetched 931 kB in 1s (1419 kB/s)
      2023-11-28 19:56:51,985:ERROR:root:stderr: 
      WARNING: apt does not have a stable CLI interface. Use with caution in scripts.
      
      E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/s/squid/squid-common_4.10-1ubuntu1.7_all.deb  404  Not Found [IP: 104.198.11.147 80]
      E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/s/squid/squid_4.10-1ubuntu1.7_amd64.deb  404  Not Found [IP: 104.198.11.147 80]
      E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```
Untested fix could be:
```diff
--- a/features/ubuntu_pro.feature
+++ b/features/ubuntu_pro.feature
@@ -3,6 +3,7 @@ Feature: Command behaviour when auto-attached in an ubuntu PRO image
     Scenario Outline: Proxy auto-attach on a cloud Ubuntu Pro machine
         Given a `<release>` `<machine_type>` machine with ubuntu-advantage-tools installed
         Given a `focal` `<machine_type>` machine named `proxy` with ingress ports `3389`
+        When I run `apt-get update` `with sudo` on the `proxy` machine
         When I run `apt install squid -y` `with sudo` on the `proxy` machine
         And I add this text on `/etc/squid/squid.conf` on `proxy` above `http_access deny all`:
             """
```

But it's better to fix it in the `Given a` code, so it only has to be done once, unless other tests rely on NOT having apt-get update run before.

**Expected behavior**

Installing packages should not fail with a 404 error, that's indicative of an out-of-date repository information and the need to run apt-get update.

**Current behavior**
Installing packages can fail.

**To Reproduce**
Run the test suite with an old-enough lxd image. Example: https://github.com/canonical/ubuntu-pro-client/actions/runs/7023435516/job/19110296166?pr=2794

**System information:**

- Pro Client version: current main

**Comment by panlinux** [comment]
Here is a jenkins failure with what looks like this problem: https://jenkins.canonical.com/server-team/view/pro/view/git/job/pro-client-integration-git-focal-lxd-vm/48/console
```
14:57:11        2024-01-01 17:57:05,868:ERROR:root:stderr: E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/l/linux-meta/linux-image-generic_5.4.0.167.164_amd64.deb  404  Not Found [IP: 185.125.190.36 80]
14:57:11        E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/l/linux-meta/linux-headers-generic_5.4.0.167.164_amd64.deb  404  Not Found [IP: 185.125.190.36 80]
14:57:11        E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/l/linux-meta/linux-generic_5.4.0.167.164_amd64.deb  404  Not Found [IP: 185.125.190.36 80]
14:57:11        E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
14:57:11        2024-01-01 17:57:05,868:WARNING:root:STEP FAILED. Collecting logs.
```

**Comment by renanrodrigo** [comment]
solved in #2918
